Allow the user to enter package tracking details mark IV: UI

ctarda opened the pull request:
Implements part of #689 

This is far from done, but I am stuck fetching the groups of tracking providers. I can not figure out why the results controller does not contain the expected data structure.

To visualise the current status:
- Navigate to an order that has not been fulfilled.
- Tap Fulfill Order > Add Tracking
- Tap Shipping Provider


Update release notes:

- [ ] If there are user facing changes, I have added an item to `RELEASE-NOTES.txt`.

mindgraffiti commented:
Hey @ctarda 👋 

So if you don't have a database viewer application yet, I would recommend downloading [DB Browser for SQLite](https://sqlitebrowser.org/). What I did to troubleshoot the problem is build the project, navigate to the screen in question, then go check out the database. To do that, I use this handy little utility, [iSimulator](https://github.com/wigl/iSimulator) that @bummytime showed me. In iSimulator, you can jump directly to your app build in the finder, then go to Sandbox > Documents > WooCommerce.sqlite. If you have the DB browser, all you have to do is double-click the .sqlite file to open it.

What I see are two stores. One for the list of shipping providers, that you are fetching in your results controller, <StorageShipmentTrackingProviderGroup>. The other store you need is <StorageShipmentTrackingProvider>. The StorageShipmentTrackingProvider contains a **group number** (which you'll have to associate with your StorageShipmentTrackingProviderGroup's Names somehow), provider **name** (such as Canada Post, DHL, etc.), and **URL**. So you are on the right track. You'll need 2 fetch results controllers, and then will need to coordinate the data from there. Or some magic in Yosemite, so that the data gets correlated before you reach a view model / view controller.

ctarda commented:
Thanks a lot @bummytime and @mindgraffiti ! I can finally present the list of providers! 🎉

ctarda commented:
👋 @bummytime @mindgraffiti 

This is becoming (has already become) too big, so I think it is better if we start looking into it piece by piece.

As it is now, it is possible to add shipment tracking manually, with some limitations:
- It is not possible to add custom tracking providers 
- There is no logic to pre-populate the form with the last shipment provider that was used.
- The "Delete Tracking" button does not work.

On the bright side, the UI is hidden behind a feature flag, so we can iterate to add all that is missing in subsequent PRs.

Thanks in advance for your time, this has become much larger that I would have liked.

ctarda commented:
This is how it looks now. 

![Screen Recording 2019-04-08 at 5 55 32 PM mov](https://user-images.githubusercontent.com/2722505/55716710-4092ce80-5a2a-11e9-95e5-34a864eb830d.gif)

I think it might be a good idea to defer the design review until we have given at least one round of review to the code, but please feel free to add the necessary tag to the PR!

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/ManualTrackingViewController.swift:
```diff
@@ -0,0 +1,496 @@
+    }
+
+    func configureBackButton() {
+        // Don't show the About title in the next-view's back button
```
Remove the word "About"

mindgraffiti commented on WooCommerce/Classes/ViewModels/ShippingProvidersViewModel.swift:
```diff
@@ -0,0 +1,55 @@
+        self.orderID = orderID
+        fetchGroups()
+    }
+
```
Needs documentation comment. something like:

```
/// Loads shipment tracking groups
///
```

mindgraffiti commented on WooCommerce/Classes/ViewModels/ShippingProvidersViewModel.swift:
```diff
@@ -0,0 +1,55 @@
+    }
+
+    private func presentNotice(_ error: Error) {
+        print("==== present error notice ====")
```
Add the error notice here. Maybe also log it? `DDLogError(error.localizedDescription)`

mindgraffiti commented on WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift:
```diff
@@ -0,0 +1,190 @@
+        get {
+            return shipmentTracking?.trackingNumber
+        }
+        set {}
```
This should be deleted or have something set inside of it.

mindgraffiti commented on WooCommerce/Classes/ViewModels/ManualTrackingViewModel.swift:
```diff
@@ -0,0 +1,190 @@
+        get {
+            return shipmentTracking?.dateShipped ?? Date()
+        }
+        set {}
```
Same here.

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/Cells/EditableOrderTrackingTableViewCell.swift:
```diff
@@ -0,0 +1,122 @@
+//        actionButton.accessibilityTraits = .button
+//        actionButton.accessibilityHint = NSLocalizedString("Tracks a shipment.",
+//                                                           comment:
+//            "Accessibility hint for Track Package button in Order details screen")
```
Uncomment code or delete it

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/FulfillViewController.swift:
```diff
@@ -82,7 +107,7 @@ private extension FulfillViewController {
     func setupTableView() {
         tableView.tableFooterView = actionView
-        tableView.allowsSelection = false
+        //tableView.allowsSelection = false
```
Delete this

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/FulfillViewController.swift:
```diff
@@ -244,14 +270,16 @@ private extension FulfillViewController {
     /// Setup a given UITableViewCell instance to actually display the specified Row's Payload.
     ///
-    func setup(cell: UITableViewCell, for row: Row) {
+    func setup(cell: UITableViewCell, for row: Row, at: IndexPath) {
```
Style nitpick. `at indexPath: IndexPath)`

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/FulfillViewController.swift:
```diff
@@ -244,14 +270,16 @@ private extension FulfillViewController {
         case .address(let shipping):
             setupAddressCell(cell, with: shipping)
+        case .tracking:
+            setupTrackingCell(cell, at: at)
```
Style nitpick. `setupTracking(cell, at: indexPath)` (I know, the rest of them should not say "Cell" as part of the method names either 😄)

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/FulfillViewController.swift:
```diff
@@ -330,10 +394,135 @@ extension FulfillViewController: UITableViewDelegate {
+            let viewModel = AddTrackingViewModel(siteID: order.siteID, orderID: order.orderID)
+            let addTracking = ManualTrackingViewController(viewModel: viewModel)
+            let navController = WooNavigationController(rootViewController: addTracking)
+            present(navController, animated: true, completion: nil)
```
Style nitpick: add a new line before the next case, to give each block of code some vertical breathing room.

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/ManualTrackingViewController.swift:
```diff
@@ -0,0 +1,496 @@
+    func displayDatePicker(at indexPath: IndexPath) {
+        datePickerVisible = true
+
+        reload()
```
What do you think about this [type of presentation](https://medium.com/@tharanit99/how-to-implement-a-inline-date-picker-in-ios-with-swift-4-9f8274460dbc) of the date picker, instead of reloading the table to make it appear? 

(The other issue with reloading the table data is that the tracking number disappears after filling it in and then tapping the date picker.)

ctarda replied:
I tried something similar (reloading just the last row) but it did crash for me. 🤔 most likely because I wasn't doing the reload within a `beginUpdates` and `endUpdates` block. I'll give it another try. FWIW the tracking number does not disappear anymore 😉

mindgraffiti replied:
>the tracking number does not disappear anymore

Hooray 😀

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/ManualTrackingViewController.swift:
```diff
@@ -0,0 +1,496 @@
+                                                                 orderID: orderID,
+                                                                 trackingID: trackingID) { [weak self] error in
+                                                                    if let error = error {
+                                                                        //track error ib Tracks
```
Does this comment mean: 

// TODO: Send error to Tracks

?

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/ManualTrackingViewController.swift:
```diff
@@ -0,0 +1,496 @@
+                                                           trackingNumber: trackingNumber) { [weak self] error in
+
+                                                            if let error = error {
+                                                                //track error ib Tracks
```
Is this supposed to be:

```// TODO: Send this error to Tracks```

?

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/ManualTrackingViewController.swift:
```diff
@@ -0,0 +1,496 @@
+// MARK: - Error handling
+//
+private extension ManualTrackingViewController {
+    /// Displays the `Unable to Add tracking` Notice.
```
excellent documentation 🙂

mindgraffiti commented on WooCommerce/Classes/ViewRelated/Orders/ShippingProvidersViewController.swift:
```diff
@@ -0,0 +1,153 @@
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        configureNavigation()
+        configureSearchController()
```
I suggest deleting the search bar methods until it's time to implement those. What do you think about implementing this part in a separate PR?

ctarda commented:
Thanks @bummytime and @mindgraffiti for looking at this behemoth.

In retrospect, I should have broken this down in smaller tasks as soon as I realised it was becoming too  big. Lesson learnt.

I have pushed an update that I think implements all of your feedback:
- Removed any search components from the list of shipping providers
- Removed launching the UI to delete a tracking (although I left the code that corresponds to that particular view model, because it has already some review comments)

Also, I have added the following issues to track what is not covered by this PR:
- [ ] #846
- [ ] #847 
- [ ] #848 
- [ ] #849 
- [ ] #850 
- [ ] #851 

Those issues seem to be granular enough to make the upcoming PRs (and reviews) shorter and easier.

mindgraffiti commented:
I did notice that this PR doesn't detect the Shipment Tracking extension. That has been added as a separate ticket, #852. Changes look great!

:shipit:

ctarda commented:
Thanks folks! Merging! 😅
